ui/ducks/send/send.js

```diff
@@ -0,0 +1,435 @@
+    },
+    updateSendTo: (state, action) => {
+      state.to = action.payload.to;
+      state.toNickname = action.payload.nickname;
+    },
+    updateSendAmount: (state, action) => {
+      state.amount = action.payload;
+    },
+    setMaxModeTo: (state, action) => {
+      state.maxModeOn = action.payload;
+    },
+    setEditingTransactionId: (state, action) => {
+      state.editingTransactionId = action.payload;
+    },
+    setSendFrom: (state, action) => {
+      state.from = action.payload;
+    },
+    updateSendToken: (state, action) => {
+      state.token = action.payload;
+      if (state.editingTransactionId && !state.token) {
+        const unapprovedTx =
+          state?.unapprovedTxs?.[state.editingTransactionId] || {};
+        const txParams = unapprovedTx.txParams || {};
+        state.tokenBalance = null;
+        state.balance = '0';
+        state.from = unapprovedTx.from ?? '';
+        txParams.data = '';
+      }
+    },
+    updateSendEnsResolution: (state, action) => {
+      state.ensResolution = action.payload;
+      state.ensResolutionError = '';
+    },
+    updateSendEnsResolutionError: (state, action) => {
+      state.ensResolution = null;
+      state.ensResolutionError = action.payload;
+    },
+    resetSendState: () => initialState,
+    gasLoadingStarted: (state) => {
+      state.gasIsLoading = true;
+    },
+    gasLoadingFinished: (state) => {
+      state.gasIsLoading = false;
+    },
+  },
+});
+
+const { actions, reducer } = slice;
+
+export default reducer;
+
+const {
+  updateSendErrors,
+  showGasButtonGroup,
+  hideGasButtonGroup,
+  setGasLimit,
+  setGasPrice,
+  setSendTokenBalance,
+  updateSendHexData,
+  updateSendTo,
+  updateSendAmount,
+  setMaxModeTo,
+  updateSendToken,
+  updateSendEnsResolution,
+  updateSendEnsResolutionError,
+  resetSendState,
+  gasLoadingStarted,
+  gasLoadingFinished,
+  setEditingTransactionId,
+  setSendFrom,
+} = actions;
+
+export {
+  updateSendErrors,
+  showGasButtonGroup,
+  hideGasButtonGroup,
+  setGasLimit,
+  setGasPrice,
+  setSendTokenBalance,
+  updateSendHexData,
+  updateSendTo,
+  updateSendAmount,
+  setMaxModeTo,
+  updateSendToken,
+  updateSendEnsResolution,
+  updateSendEnsResolutionError,
+  setSendFrom,
+  setEditingTransactionId,
+  resetSendState,
+};
+
+export function updateSendTokenBalance({ sendToken, tokenContract, address }) {
+  return (dispatch) => {
+    const tokenBalancePromise = tokenContract
+      ? tokenContract.balanceOf(address)
+      : Promise.resolve();
+    return tokenBalancePromise
+      .then((usersToken) => {
+        if (usersToken) {
+          const newTokenBalance = calcTokenBalance({ sendToken, usersToken });
+          dispatch(setSendTokenBalance(newTokenBalance));
+        }
+      })
+      .catch((err) => {
+        log.error(err);
+        updateSendErrors({ tokenBalance: 'tokenBalanceError' });
+      });
+  };
+}
+
+async function estimateGasLimitForSend({
```

Comment:
It would be nice to share the logic here with the background, where we also have a function to set default gas parameters and measure the gas limit.

Reply:
Didn't go this far yet, but will in a future PR